pkg/apis/submariner.io/v1/endpoint.go

```diff
@@ -102,3 +103,50 @@ func (ep *EndpointSpec) hasSameBackendConfig(other *EndpointSpec) bool {
 
 	return equality.Semantic.DeepEqual(ep.BackendConfig, other.BackendConfig)
 }
+
+func fromArrayOrString(a []string, s string) []string {
+	if len(a) > 0 || s == "" {
+		return a
+	}
+
+	return []string{s}
+}
+
+func toIPArrayAndV4String(ips []string) ([]string, string) {
+	v4IP := ""
+	ipArray := make([]string, len(ips))
+
+	for i := range ips {
+		ipArray[i] = ips[i]
+
+		if k8snet.IsIPv4String(ips[i]) {
+			v4IP = ips[i]
+		}
+	}
+
+	return ipArray, v4IP
+}
+
+func (ep *EndpointSpec) GetHealthCheckIPs() []string {
+	return fromArrayOrString(ep.HealthCheckIPs, ep.HealthCheckIP)
+}
+
+func (ep *EndpointSpec) SetHealthCheckIPs(ips ...string) {
+	ep.HealthCheckIPs, ep.HealthCheckIP = toIPArrayAndV4String(ips)
+}
+
+func (ep *EndpointSpec) GetPublicIPs() []string {
+	return fromArrayOrString(ep.PublicIPs, ep.PublicIP)
+}
+
+func (ep *EndpointSpec) SetPublicIPs(ips ...string) {
```

Comment:
overall code LGTM,
I have one minor comment,  since some of the fields (e.g Public IP) can be discovered at different timing maybe it makes sense to replace (or add new API) SetxxxIPs with UpdatexxIP (Update should call GetxxxIPs and update it if needed ) instead of having this logic in clients side

Reply:
Sorry I'm not sure I follow what you mean...

Reply:
I'm sorry, but I'll try to explain it by example ... 

Assuming the following use case for dual-stack case: 
A. public IPV4 address successfully discovered on t1
B. public IPV6 address only discovered on t2 

I thought it might be useful to have Update API (that will add  IP if it doesn't exist, we can use it in step B for example).

But on second thought since there is only single client (gateway) that will call this APIs we can have this logic at the gateway side.

Reply:
I see - there may be cases where we need to set IPs atomically or piecemeal. Either way we can add/modify the helper functions as needed. The main point of these helpers is to hide the singular fields.

Reply:
This highlights that there’s a difference between the way the data is stored and how it’s used. Externally, the API could be `Set…IPv4`/`Set…IPv6` (with corresponding getters); that would make it explicit that the data structure only stores one IPv4 address and one IPv6 address.

(Incidentally, there’s currently nothing stopping a caller from storing two IPv4 addresses or two IPv6 addresses :wink:.)

Reply:
OK I changed the getters to retrieve the singular IP for an IP family. The family is passed as a parameter as, after looking at usages, I think that's going to be more natural/easier rather than separate functions for each family. The setters add the address if not already present to ensure there's only one per family.